compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1266,7 +1266,8 @@ void handleDeclaredVarInForeach(BLangVariable variable, BType rhsType, SymbolEnv
                 break;
             case TUPLE_VARIABLE:
                 BLangTupleVariable tupleVariable = (BLangTupleVariable) variable;
-                if (TypeTags.TUPLE != rhsType.tag && TypeTags.UNION != rhsType.tag) {
+                if ((TypeTags.TUPLE != rhsType.tag && TypeTags.ARRAY != rhsType.tag && TypeTags.UNION != rhsType.tag) ||
+                        (variable.isDeclaredWithVar && !types.isSubTypeOfBaseType(rhsType, TypeTags.TUPLE))) {
```

Comment:
Shouldn't we check `!types.isSubTypeOfBaseType(rhsType, TypeTags.ARRAY)` also here?

Reply:
If rhsType is not a tuple type, an array type, or a union with all members as tuple/array, the `isSubTypeOfBaseType` check anyway be false. We don't need to do the `isSubTypeOfBaseType` check for array type again then?

Reply:
Ah, yes!